# Practice Case MongoDB
### This Code Belongs to Dwilaras Athina

## Instruction
Connect cluster pymongo pada link berikut (pilih salah satu yang bisa) :


*   mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority
*   mongodb+srv://mongodb-intro:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true
&w=majority


1. Membuat collection baru dengan nama clean_movies yang sama persis dengan movies collection
pada database sample_mflix dengan memakai collection movies_intial (projecting_queries pymongo)

2. Validasi collection yang telah dibuat dengan parameter sbb :
 

*   Semua document pada clean_movies dan movie sama
*   Banyak document pada clean_movies dan movie sama
*   Semua fields pada clean_movies ada pada movie
*   Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

_Sertakan scriptnya dengan format judul script : practice1_nama.ipynb_


================================================================================================

## Answer

Table of Content :

<ul>

<li><a href="#no1">1. Install Module PyMongo & Import Libraries</a></li>
  
<li><a href="#no2">2. Initiate Connection</a></li>
  
<li><a href="#no3">3. Show All Database in Cluster</a></li>
  
<li><a href="#no4">4. Show All Collection in Used Database</a></li>
  
<li><a href="#no5">5. Making a New Collection</a></li>
  
<li><a href="#no6">6. Validation</a></li>
  
<li><a href="#no7">7. Show MongoDB Compas for New Collection : clean_movies_dwilaras</a></li>
  
</ul>

<a id='no1'></a>
### 1. Install Module PyMongo & Import Libraries

In [69]:
! pip install pymongo 
! pip install pymongo[srv]
! pip install pprint

In [0]:
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

<a id='#no2'></a>
### 2. Initiate Connection 

In [0]:
client = MongoClient ("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority")

In [54]:
print(client)

MongoClient(host=['cluster0-shard-00-02-nnbxe.gcp.mongodb.net:27017', 'cluster0-shard-00-00-nnbxe.gcp.mongodb.net:27017', 'cluster0-shard-00-01-nnbxe.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


<a id='no3'></a>
### 3. Show All Database in Cluster

In [55]:
print(client.list_database_names())

['sample_mflix', 'admin', 'local']


Database in Cluster

1.   sample_mflix
2. admin
3. local



<a id='no4'></a>
### 4. Show All Collection in Used Database

In [0]:
# Database client:sample_mflix
db = client['sample_mflix']
# Col is movies_initial from Database
# Col2 is movies from Database
col = db.movies_initial
col2 = db.movies

In [57]:
# show all collection in used db
print(db.list_collection_names())

['clean_nama_kamu', 'clean_movies_nurlailiis', 'clean_movies_devita', 'movies_initial', 'clean_movies_frans', 'clean_movies_tara', 'clean_movies_tunggal', 'clean_movies', 'users', 'clean_movies_nurrimah', 'movies_testtopik', 'movies_scratch', 'clean_movies_topik', 'movies', 'clean_movies_afifah', 'clean_movies_andreas', 'clean_movies_indra_1']


Collection in Database :


1. clean_nama_kamu
2. clean_movies_nurlailiis
3. clean_movies_devita
4. movies_initial
5. clean_movies_frans
6. clean_movies_tara
7. clean_movies_tunggal
8. clean_movies
9. users
10. clean_movies_nurrimah
11. movies_testtopik
12. movies_scratch
13. clean_movies_topik
14. movies
15. clean_movies_afifah
16. clean_movies_andreas
17. clean_movies_indra_1

<a id='no5'></a>
### 5. Making a New Collection
> Instruction : Membuat collection baru dengan nama clean_movies yang sama persis dengan movies collection
pada database sample_mflix dengan memakai collection movies_intial (projecting_queries pymongo)

#### 5.1 Showing Keys in Movies Initial & Movies Collection

In [0]:
# show keys MOVIES INITIAL from one of document (random)
x = col.find_one()
print(x.keys())

dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type'])


Keys in movies_initial

1. _id
2. imdbID
3. title
4. year
5. rating
6. runtime
7. genre
8. released
9. director
10. writer
11. cast
12. metacritic
13. imdbRating
14. imdbVotes
15. poster
16. plot
17. fullplot
18. language
19. country
20. awards
21. lastupdated
22. type



In [0]:
# how to show all keys MOVIES INITIAL in one vertical list
a = [list(doc.keys()) for doc in col.find({})]
a = set([j for i in a for j in i])
a

{'_id',
 'awards',
 'cast',
 'country',
 'director',
 'fullplot',
 'genre',
 'imdbID',
 'imdbRating',
 'imdbVotes',
 'language',
 'lastupdated',
 'metacritic',
 'plot',
 'poster',
 'rating',
 'released',
 'runtime',
 'title',
 'type',
 'writer',
 'year'}

In [0]:
# show keys in MOVIES from one of document (random)
x = col2.find_one()
print(x.keys())

dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'countries', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'type', 'tomatoes'])


Keys in Movies:
1. _id
2. plot
3. genres
4. runtime
5. cast
6. num_mflix_comments
7. poster
8. title
9. fullplot
10. countries
11. released
12. directors
13. writers
14. awards
15. lastupdated
16. year
17. imdb
18. type
19. tomatoes

In [0]:
# how to show all keys MOVIES in one vertical list
b = [list(doc.keys()) for doc in col2.find({})]
b=set([j for i in b for j in i])
b

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

**Note:**


*   Jika kita menggunakan find_one(), maka keys yang kita dapatkan akan random dari salah satu dokumen.
*   Sedangkan, jika kita menggunakan kode dibawah ini, kita dapat mendapatkan keys pada seluruh dokumen. 



> b = [list(doc.keys()) for doc in col2.find({})]
>     
> b = set([j for i in b for j in i])
>
> print(b)






#### 5.2 Showing Collection in DataFrame Format

In [0]:
# movies_initial collection 
import pandas as pd
pd.DataFrame(list(col.find().limit(4)))

_id  awards  ... writer  year
0  5daffb1b986c4adedb754af9          ...         1894
1  5daffb1b986c4adedb754afa  1 win.  ...         1893
2  5daffb1b986c4adedb754afb          ...         1892
3  5daffb1b986c4adedb754afc          ...         1894

[4 rows x 22 columns]

In [0]:
# movies collection
pd.DataFrame(list(col2.find().limit(3)))

_id  ...  year
0  573a1390f29313caabcd5a93  ...  1916
1  573a1390f29313caabcd60e4  ...  1917
2  573a1391f29313caabcd71f5  ...  1921

[3 rows x 21 columns]

**Note:** 


*   Dengan menggunakan dataframe format, kita dapat secara langsung membandingkan value setiap key dari masing-masing collection.
*   Kita akan menggunakan detail dari masing-masing value untuk membentuk newcollection yang bernama clean_movies_dwilaras. 



#### 5.3 Filtering the Movies in Movies Collection Only

> Sekarang, saya ingin filter movies yang terdapat pada movies_initial (col). Yang akan saya ambil hanyalah movies yang sama persis dengan yang berada pada movies collection (col2).

In [58]:
imdb= list (col2.find({},{'_id':0, 'imdb':1}))
imdb_list=[]
print(len(imdb))
for i in imdb:
  imdb_list.append(list(i.values())[0])

imdb_list[:5]

23541


[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

**Note:**
> Dapat terlihat, jika yang kita inginkan adalah movie dari movies collection, maka terdapat 23541 movies.
>   
> 23541 movies tersebut akan di masukkan (di input) kedalam clean_movies_dwilaras.

#### 5.4 Make a Pipeline

In [104]:
# make a pipeline for new collection, named : "clean_movies_dwilaras"

pipeline = [
        {
        '$project': {
            # this code is to include the keys for a new collection on database
            # 1 for include , 0 for exclude the keys,
            
            '_id':1,
            'title': 1,
            'year': 1,
            'cast': 1,
            'fullplot':1,
            'lastupdated': 1,
            'plot':1,
            'poster': 1,
            'type':1,
            'metacritic':1,
            'num_mflix_comments':"",
            
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            'released': { '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]},
                                      ", 0, 0)" ] },
                    
            # here we split some keys into an array value, renaming as well 
            'directors': {'$split': ["$director", ", "]},
            'cast': {'$split': ["$cast", ", "]},
            'writers': {'$split': ["$writer", ", "]},
            'genres': {'$split': ["$genre", ", "]},
            'countries': {'$split': ["$country", ", "]},
            'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            
            # renaming
            'languages':"$language",
            'rated':"$rating",
   
            
            # adding embeded documents
            # grouping the several keys into embeded documents by one keys 
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },
            
            'tomatoes' :{
                'lastUpdated':"",
                'viewer':{'meter':"",'numReviews':"",'rating':""},
                'dvd':'',
                'fresh':'',
                'rotten':'',
                'production':''
                
            }
 
        }
    },
    
    {
        "$match":{"imdb":{"$in":[dict(sorted(imdb_list[x].items(), key=lambda x:x[0])) for x in range(len(imdb_list))]}}
    },
  
    {
        # naming the new collection as output 
        '$out': "clean_movies_dwilaras"
    }
]

pprint.pprint(list(client.sample_mflix.movies_initial.aggregate(pipeline)))

[]


**Note:**
Ada beberapa catatan untuk pipeline yang saya buat. Mungkin, ini akan memudahkan anda sebagai pembaca untuk mempelajarinya atau untuk mengembangkannya. 


1.   Untuk mendapatkan hasil yang sama, pertama-tama kita harus pastikan keys apa saja yang berada pada movies collection. 
2.   Kita dapat membuat penamaan baru pada nama keys yang berbeda. Contohnya, pada sumber data movies_initial terdapat language, sedangkan pada movies, key nya berubah nama menjadi languages. Kita dapat mengganti dengan cara seperti dibawah ini :
> 'languages':"$language"

3. Kita dapat _splitting_ keys menjadi array value. Contohnya, pada sumber data movies_initial terdapat key director, sedangkan pada movies, key director berubah bentuk menjadi array. Kita dapat mengubahnya dengan cara seperti dibawah ini :
> 'directors': {"$split": ["$director", ", "]}

4. Kita dapat membuat embeded document. Contohnya, pada sumber data movies_initial terdapat key imdbID, imdbRating, dan imdbVotes. Sedangkan pada movies, ketiga keys tersebut, masuk kedalam satu key yang bernama 'imdb'. Kita dapat mengubahnya dengan cara seperti dibawah ini :
>   'imdb': {
              'id': "$imdbID",
             'rating':"$imdbRating",
               'votes': "$imdbVotes" }
               
5. Kita dapat membuat keys baru walaupun dari sumber data tidak ada keys tersebut. 
> 'tomatoes' :{
                'lastUpdated':"",
                'viewer':
                {'numReviews':"",'rating':""}
                
                }

6. Kita dapat meng-input (mengisi) data yang hanya kita inginkan, dengan cara filtering di awal, setelah itu, kita buatkan sebuah matching. Code dapat dituliskan seperti dibawah ini :
> "$match":{"imdb":{"$in":
    [dict(sorted(imdb_list[x].items(), 
    key=lambda x:x[0])) for x in 
    range(len(imdb_list))]}}

7. Buat Nama dari collection baru dengan cara berikut ini : 
> '$out': "clean_movies_dwilaras"




#### 5.5 Print One of clean_movies_dwilaras Document

In [105]:
pprint.pprint(db.clean_movies_dwilaras.find_one())

{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'languages': '',
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'metacritic': '',
 'num_mflix_comments': '',
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'poster

#### 5.6 Make a Name for New Collection "clean_movies_dwilaras"

In [0]:
col_dwilaras=db.clean_movies_dwilaras

**Note:**
> Penamaan colection ini akan memudahkan kita saat memanggil _db.clean_movies_dwilaras_ pada analisis berikutnya. 

<a id='no6'></a>
### 6. Validation

#### 6.1 Semua document dan jumlah document pada clean_movies dan movie sama

In [67]:
#Check jumlah document pada collection 'clean_movies_dwilaras'
len(list(col_dwilaras.find()))

23539

In [71]:
#Check jumlah document pada collection 'movies'
len(list(col2.find()))

23541

**Note:**
*   Ternyata setelah diperiksa, jumlah document pada clean_movies_dwilaras ada 23.539 documents. Sementara, jumlah document pada movies ada 23.541 documents.
*   Mari kita lihat apakah yang membuatnya berbeda. 



In [0]:
# Get different document(s)
# Grouping and sortcount based on imdb.
dwilarasgroup = list(db['clean_movies_dwilaras'].aggregate([{'$sortByCount':'$imdb'}]))
moviesgroup = list(db['movies'].aggregate([{'$sortByCount':'$imdb'}]))

In [73]:
#Get the IMDB which different between dwilarasgroup and moviesgroup
for i in moviesgroup:
  if i not in dwilarasgroup:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


**Note:**


*   Didapatkan IMDB Id: 5 yang berbeda. 
*   Selanjutnya, tampilkan dokumen dengan IMDB id: 5



In [106]:
# Show document that imdb id 5 in clean_movies_dwilaras
pprint.pprint(list(col_dwilaras.find({"imdb.id":{"$in":[5]}})))

[{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'languages': '',
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'metacritic': '',
  'num_mflix_comments': '',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer

In [91]:
# Show document that imdb id 5 in movies
pprint.pprint(list(col2.find({"imdb.id":{"$in":[5]}})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

**Note:**
 

*   Setelah dilihat, ternyata perbedaannya adalah pada **title**. Pada movies, title nya adalah Blacksmith Scene dan Blacksmith Scene II. Sedangkan pada clean_movies_dwilaras title nya hanya Blacksmith Scene. Hal tersebut akan diabaikan, karena ada kemungkinan hanyalah salah input. Bisa juga untuk di drop, jika tidak diinginkan movies dengan judul Blacksmith Scene II. 
*   Perbedaan juga ditemukan pada **genres**. Untuk movies, ada yang bergenre Short, ada juga yang bergenre Short dan Long. Sedangkan untuk clean_movies_dwilaras, hanya ada genre Short. 



#### 6.2 Semua fields dan banyak fields pada clean_movies dan movie sama

In [0]:
# how to show all keys MOVIES in one vertical list
b = [list(doc.keys()) for doc in col2.find({})]
b = set([j for i in b for j in i])
b

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [86]:
# count all keys in movies
print(len(b))

22


In [85]:
# how to show all keys clean_movies_dwilaras in one vertical list
c = [list(doc.keys()) for doc in col_dwilaras.find({})]
c = set([j for i in c for j in i])
c

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [87]:
# count all keys in clean_movies_dwilaras
print(len(c))

22


**Note:**
> Sudah dipastikan bahwa jumlah keys pada clean_movies_dwilaras dan movies sama, yaitu terdapat 22 keys. 

#### 6.3 Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

> Saya akan coba cek untuk tiga movies pada clean_movies_dwilaras dan movies. Hal tersebut sudah merepresentasikan all-movies. 

#### 6.3.1 Movies dengan Title : Gertie the Dinosaur

In [107]:
pprint.pprint(list(col_dwilaras.find({'title': 'Gertie the Dinosaur'})))
print ('\n ----------------------------------------------------------- \n')
pprint.pprint(list(col2.find({'title': 'Gertie the Dinosaur'})))

[{'_id': ObjectId('5daf3e85ab0f2781ea55d5ec'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Winsor McCay', 'George McManus', 'Roy L. McCardell'],
  'countries': ['USA'],
  'directors': ['Winsor McCay'],
  'fullplot': 'Winsor Z. McCay bets another cartoonist that he can animate a '
              'dinosaur. So he draws a big friendly herbivore called Gertie. '
              'Then he get into his own picture. Gertie walks through the '
              'picture, eats a tree, meets her creator, and takes him '
              'carefully on her back for a ride.',
  'genres': ['Animation', 'Short', 'Comedy'],
  'imdb': {'id': 4008, 'rating': 7.3, 'votes': 1837},
  'languages': 'English',
  'lastupdated': '2015-08-18 01:03:15.313000000',
  'metacritic': '',
  'num_mflix_comments': '',
  'plot': 'The cartoonist, Winsor McCay, brings the Dinosaurus back to life in '
          'the figure of his latest creation, Gertie the Dinosaur.',
  'poster': 'http://ia.media-imdb

#### 6.3.2 Movies dengan Title : The Cheat

In [108]:
pprint.pprint(list(col_dwilaras.find({'title': 'The Cheat'})))
print ('\n ----------------------------------------------------------- \n')
pprint.pprint(list(col2.find({'title': 'The Cheat'})))

[{'_id': ObjectId('5daf3e85ab0f2781ea55d604'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Fannie Ward', 'Sessue Hayakawa', 'Jack Dean', 'James Neill'],
  'countries': ['USA'],
  'directors': ['Cecil B. DeMille'],
  'fullplot': 'Edith Hardy uses charity funds for Wall Street investments in '
              'hopes of buying some new gowns. She loses all the money and '
              'borrows from wealthy oriental Tori. When her husband gives her '
              "the amount she borrowed, Tori won't take it back, branding her "
              'shoulder with a Japanese sign of his ownership. She shoots him. '
              'Her husband takes the blame. In court Edith reveals all to an '
              'angry mob.',
  'genres': ['Drama'],
  'imdb': {'id': 5078, 'rating': 6.5, 'votes': 1660},
  'languages': 'English',
  'lastupdated': '2015-08-31 00:41:20.670000000',
  'metacritic': '',
  'num_mflix_comments': '',
  'plot': "A venal, spoiled stockbroker's wife 

#### 6.3.3 Movies dengan Title : Intolerance: Love's Struggle Throughout the Ages

In [109]:
pprint.pprint(list(col_dwilaras.find({'title': "Intolerance: Love's Struggle Throughout the Ages"})))
print ('\n ----------------------------------------------------------- \n')
pprint.pprint(list(col2.find({'title': "Intolerance: Love's Struggle Throughout the Ages"})))

[{'_id': ObjectId('5daf3e85ab0f2781ea55d628'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Lillian Gish',
           'Spottiswoode Aitken',
           'Mary Alden',
           'Frank Bennett'],
  'countries': ['USA'],
  'directors': ['D.W. Griffith'],
  'fullplot': 'Intolerance and its terrible effects are examined in four '
              'historical eras. In ancient Babylon, a mountain girl is caught '
              "up in the religious rivalry that leads to the city's downfall. "
              'In Judea, the hypocritical Pharisees condemn Jesus Christ. In '
              "1572 Paris, unaware of the impending St. Bartholomew's Day "
              'Massacre, two young Huguenots prepare for marriage. Finally, in '
              'modern America, social reformers destroy the lives of a young '
              'woman and her beloved.',
  'genres': ['Drama', 'History'],
  'imdb': {'id': 6864, 'rating': 8, 'votes': 9880},
  'languages': '',
  'lastupdated': '2

**Note:**
*   Setelah dilihat dari movie **Gertie the Dinosaur**, **The Cheat**, dan  **Intolerance: Love's Struggle Throughout the Ages** urutannya sudah sama. 
*   Untuk isi (values) semua sudah sama. Kecuali 'num_mflix_comments' dan di tomatoes. Hal tersebut dikarenakan sumber data initial_movies (col) tidak memiliki keys tersebut. Saya hanya menampilkan keys saja, tanpa ada isi (values).
*   Sehingga, dapat dipastikan semua movies pada clean_movies_dwilaras isi dan urutannya sudah sama dengan movies.




<a id='no7'></a>
### 7. Show MongoDB Compas for New Collection : clean_movies_dwilaras

![MongoDB Compass](compass.jpg)

## Conclussion

Pada Practice Case kali ini, saya telah belajar tentang bagaimana membuat collection baru sesuai dengan kriteria yang diinginkan. 

 

*   Pertama: kita harus melalakukan Install Module PyMongo dan import libraries yang diinginkan.
*   Kedua: kita buat connection dengan cluster yang kita inginkan. 
*   Ketiga: kita bisa menampilkan database pada cluster yang tersedia, dan collection yang kita inginkan pada database. 
*   Keempat: setelah kita menampilkan keys pada sumber data (movies_initial collection) dan movies collection, kita buat pipeline untuk projecting. 
*   Kita harus memastikan bahwa saat membuat pipeline sesuai dengan kebutuhan. Beberapa cara dapat kita lakukan, misalnya renaming, splitting, membuat embeded document, dan matching. 
*   Terakhir, kita dapat melakukan validating. Dengan memastikan bahwa jumlah dokumennya, keys, jumlah keys, values, dan urutannya sama semuanya. 

Sekilas, anda dapat melihat collection baru dengan nama 'clean_movies_dwilaras' dengan melihat tampilan di MongoDB Compass diatas. 


